pkg/agent/openflow/multicluster.go

```diff
@@ -0,0 +1,231 @@
+	localGatewayMAC net.HardwareAddr,
+	peerServiceCIDR net.IPNet,
+	tunnelPeer net.IP,
+	remoteGatewayIP net.IP) []binding.Flow {
+	ipProtocol := getIPProtocol(remoteGatewayIP)
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var flows []binding.Flow
+	flows = append(flows,
+		// This generates the flow to forward cross-cluster request packets based
+		// on Service ClusterIP range.
+		L3ForwardingTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
+			MatchProtocol(ipProtocol).
+			MatchDstIPNet(peerServiceCIDR).
+			Action().SetSrcMAC(localGatewayMAC).                 // Rewrite src MAC to local gateway MAC.
+			Action().SetDstMAC(GlobalVirtualMACForMulticluster). // Rewrite dst MAC to virtual Multi-cluster MAC.
+			Action().SetTunnelDst(tunnelPeer).                   // Flow based tunnel. Set tunnel destination.
+			Action().LoadRegMark(ToTunnelRegMark).
+			Action().GotoTable(L3DecTTLTable.GetID()).
+			Done(),
+		// This generates the flow to forward cross-cluster reply traffic based
+		// on Gateway IP.
+		L3ForwardingTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
+			MatchProtocol(ipProtocol).
+			MatchDstIP(remoteGatewayIP).
+			MatchCTStateRpl(true).
+			MatchCTStateTrk(true).
+			Action().SetSrcMAC(localGatewayMAC).
+			Action().SetDstMAC(GlobalVirtualMACForMulticluster).
+			Action().SetTunnelDst(tunnelPeer). // Flow based tunnel. Set tunnel destination.
+			Action().LoadRegMark(ToTunnelRegMark).
+			Action().GotoTable(L3DecTTLTable.GetID()).
+			Done(),
+	)
+	return flows
+}
+
+func (f *featureMulticluster) tunnelClassifierFlow(tunnelOFPort uint32) binding.Flow {
+	return ClassifierTable.ofTable.BuildFlow(priorityHigh).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchInPort(tunnelOFPort).
+		MatchDstMAC(GlobalVirtualMACForMulticluster).
+		Action().LoadRegMark(FromTunnelRegMark).
+		Action().LoadRegMark(RewriteMACRegMark).
+		Action().GotoStage(stageConntrackState).
+		Done()
+}
+
+func (f *featureMulticluster) outputFlow(tunnelOFPort uint32) binding.Flow {
+	return L2ForwardingOutTable.ofTable.BuildFlow(priorityHigh).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchRegMark(RewriteMACRegMark).
+		MatchRegMark(OutputToTunnelRegMark).
+		MatchInPort(tunnelOFPort).
+		Action().OutputInPort().
+		Done()
+}
+
+// snatConntrackFlows generates the flow to match the first packet of a cross-cluster connection.
+func (f *featureMulticluster) snatConntrackFlows(serviceCIDR net.IPNet, localGatewayIP net.IP) []binding.Flow {
+	var flows []binding.Flow
+	ipProtocol := getIPProtocol(localGatewayIP)
+
+	flows = append(flows,
+		// This generates the flow to match the first packet of a Service connection sourced from a local Pod or local
+		// Antrea gateway, and the Endpoint is a Service ClusterIP on remote cluster. Note that, DNAT is performed
+		// on such connection, and SNAT should also be performed on such connection to forward the packets to remote cluster.
```

Comment:
Remove "on such connection".

“to forward" -> "before sending".

Reply:
”the remote cluster"

Reply:
Updated.